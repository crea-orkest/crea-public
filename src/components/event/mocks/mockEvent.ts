import type { Event } from 'graphql/types/event'

export const mockEvent: Event = {
  id: 'id',
  title: 'name',
  image: undefined,
  locations: [{ startTime: 'DATETIME', id: 'test id' }],
  persons: [],
}
