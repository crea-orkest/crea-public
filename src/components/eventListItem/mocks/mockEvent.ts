import type { Event } from 'graphql/types/event'

export const mockEvent: Event = {
  __typename: 'ConcertRecord',
  _createdAt: '2023-09-01T13:32:36+01:00',
  _firstPublishedAt: undefined,
  _publishedAt: undefined,
  _updatedAt: '2023-08-14T09:37:24+01:00',
  id: 'id',
  title: 'name',
  image: undefined,
  locations: [{ startTime: '2023-09-01T13:32:36+01:00', id: 'test id' }],
  persons: [],
  program: [],
  url: '/some-slug',
  slug: 'some-slug',
  content: [],
}
