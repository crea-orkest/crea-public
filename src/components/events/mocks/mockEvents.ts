import type { Event } from 'graphql/types/event'

export const mockEvent: Event = {
  id: 'mock-id',
  title: 'title',
  url: '/concerten/concert-1',
  image: undefined,
  locations: [],
  persons: [],
}
