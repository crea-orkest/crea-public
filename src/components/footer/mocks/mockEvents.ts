import { type Event } from 'graphql/types/event'

export const mockEvent: Event = {
  id: 'mock-id',
  title: 'title',
  image: undefined,
  locations: [],
  persons: [],
  url: '/some-slug',
  content: [],
}
