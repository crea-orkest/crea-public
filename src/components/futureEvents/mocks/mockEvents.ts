import { type Event } from 'graphql/types/event'

export const mockEvent: Event = {
  id: 'mock-id',
  title: 'title',
  image: undefined,
  locations: [],
  persons: [],
  program: [],
  url: '/some-slug',
  content: [],
}
