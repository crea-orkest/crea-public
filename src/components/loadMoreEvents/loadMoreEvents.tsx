'use client'
import type { Event as EventType } from 'graphql/types/event'
import { EventView } from 'components/event/eventView'
import React from 'react'
import { getEvents } from 'graphql/getters/getEvents'
import { useEventsMeta } from 'hooks/useEventsMeta'
import { useIntersectionObserver } from 'hooks/useIntersectionObserver'

export interface Props {
  initialSkip: number
}

export const LoadMoreEvents = ({ initialSkip }: Props) => {
  const [skip, setSkip] = React.useState(initialSkip)
  const [events, setEvents] = React.useState<EventType[]>([])
  const ref = React.useRef<HTMLButtonElement>(null)
  const { numberOfEvents } = useEventsMeta()
  const [entry] = useIntersectionObserver({
    condition: Boolean(numberOfEvents),
    ref,
  })

  React.useEffect(() => {
    let fetching = false
    if (!numberOfEvents) return
    if (events.length + initialSkip === numberOfEvents) return
    if (!entry?.isIntersecting) return
    if (fetching) return
    if (skip >= numberOfEvents) return
    fetching = true
    // TODO: abort signal to stop fetching
    getEvents({ skip, first: 1 })
      .then(({ data }) => {
        if (!data) return
        setEvents((prev) => [...prev, ...data])
        setSkip((prev) => {
          return prev + 1
        })
      })
      .catch(() => {
        console.log('TODO: render an error')
      })
      .finally(() => {
        fetching = false
      })
  }, [
    skip,
    setSkip,
    numberOfEvents,
    entry?.isIntersecting,
    events.length,
    initialSkip,
  ])

  return (
    <div>
      {events.map((event) => {
        return <EventView key={event.id} data={event} />
      })}

      {numberOfEvents && (
        <button
          type="button"
          ref={ref}
          disabled={events.length + initialSkip === numberOfEvents}
          onClick={() =>
            console.log('TODO: fetch if no IntersectionObserver support')
          }
        >
          load more elements
        </button>
      )}

      {numberOfEvents && (
        <p>
          {/* TODO: fix this logic */}
          {events.length + initialSkip} van {numberOfEvents} concerten
        </p>
      )}
    </div>
  )
}
