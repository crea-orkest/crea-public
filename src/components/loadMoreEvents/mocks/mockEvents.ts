import { type Event } from 'graphql/types/event'

export const mockEvent: Event = {
  __typename: 'ConcertRecord',
  _createdAt: '2023-09-01T13:32:36+01:00',
  _firstPublishedAt: undefined,
  _publishedAt: undefined,
  _updatedAt: '2023-08-14T09:37:24+01:00',
  id: 'mock-id',
  title: 'title',
  slug: 'url',
  url: '/url',
  image: undefined,
  locations: [],
  persons: [],
  program: [],
}
