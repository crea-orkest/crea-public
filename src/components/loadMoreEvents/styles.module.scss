@use 'sass:map';
@use '../../styles/tokens/spacing.scss' as spacing;

.loaderText {
  display: flex;
  gap: map.get(spacing.$rem, 8);
  align-items: flex-end;
  justify-content: center;
}

.loader {
  width: 32px;
  height: 32px;
}
