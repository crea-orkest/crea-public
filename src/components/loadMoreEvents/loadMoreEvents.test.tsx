import { LoadMoreEvents } from './loadMoreEvents'
import React from 'react'
import { getEvents } from 'graphql/getters/getEvents'
import { mockEvent } from './mocks/mockEvents'
import { render } from '@testing-library/react'

jest.mock('../../graphql/getters/getEvents', () => {
  const originalModule = jest.requireActual('../../graphql/getters/getEvents')
  return {
    __esModule: true,
    ...originalModule,
    getEvents: jest.fn(),
  }
})

jest.mock('../../hooks/useEventsMeta', () => {
  const originalModule = jest.requireActual('../../hooks/useEventsMeta')
  return {
    __esModule: true,
    ...originalModule,
    useEventsMeta: () => ({ maxEvents: 2 }),
  }
})

const getEventsMock = jest.mocked(getEvents)

describe('Events component', () => {
  it('shows all the data', async () => {
    getEventsMock.mockResolvedValue({
      data: [mockEvent],
      error: undefined,
    })

    const { container } = render(<LoadMoreEvents initialSkip={1} />)

    expect(container).toMatchSnapshot()
  })
})
