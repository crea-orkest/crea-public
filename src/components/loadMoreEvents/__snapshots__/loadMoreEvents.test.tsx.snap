// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events component shows all the data 1`] = `
<div>
  <div>
    <button
      type="button"
    >
      load more elements
    </button>
    <p>
      1
       van 
      10
       concerten
    </p>
  </div>
</div>
`;
